pallets/gear/src/queue.rs

```diff
@@ -95,116 +101,171 @@ where
-                unreachable!(
-                    "`allocations_tree_len` {} is not zero, so program {destination_id:?} must have allocations",
-                    program.allocations_tree_len,
-                )
-            })
-        }).unwrap_or_default();
-
-        let actor_data = ExecutableActorData {
-            allocations,
-            code_id: program.code_hash.cast(),
-            code_exports: program.code_exports,
-            static_pages: program.static_pages,
-            gas_reservation_map: program.gas_reservation_map,
-            memory_infix: program.memory_infix,
-        };
+        log::debug!("Gas burned after CodeMetadata {:?}", context.gas_burned());
 
-        // The second step is to load instrumented binary code of the program but
-        // first its correct length should be obtained.
-        let context =
-            match core_processor::precharge_for_code_length(block_config, context, actor_data) {
-                Ok(context) => context,
-                Err(journal) => return journal,
-            };
+        let code_id = program.code_id;
 
-        // Load correct code length value.
-        let code_id = context.actor_data().code_id;
-        let code_len_bytes = T::CodeStorage::get_code_len(code_id).unwrap_or_else(|| {
-            // `Program` exists, so do code and code len.
+        // The second step is to load code metadata
+        let code_metadata = T::CodeStorage::get_code_metadata(code_id).unwrap_or_else(|| {
             let err_msg = format!(
-                "run_queue_step: failed to get code len for the existing program. \
+                "run_queue_step: failed to get code metadata for the existing program. \
                 Program id -'{destination_id:?}', Code id - '{code_id:?}'."
             );
 
             log::error!("{err_msg}");
             unreachable!("{err_msg}");
         });
 
-        // Adjust gas counters for fetching instrumented binary code.
-        let context =
-            match core_processor::precharge_for_code(block_config, context, code_len_bytes) {
-                Ok(context) => context,
-                Err(journal) => return journal,
-            };
+        if !code_metadata.exports().contains(&dispatch_kind) {
+            let (destination_id, dispatch, gas_counter, _) = context.into_parts();
 
-        // Load instrumented binary code from storage.
-        let code = T::CodeStorage::get_code(code_id).unwrap_or_else(|| {
-            // `Program` exists, so do code and code len.
-            let err_msg = format!(
-                "run_queue_step: failed to get code for the existing program. \
-                Program id -'{destination_id:?}', Code id - '{code_id:?}'."
+            return core_processor::process_success(
+                SuccessfulDispatchResultKind::Success,
+                DispatchResult::success(dispatch, destination_id, gas_counter.to_amount()),
             );
+        }
```

Comment:
Should be executed after reinstrumentation

Reply:
hmm, is it really possible for exports to change after re-instrumentation?

Reply:
added a note to rewrite in case re-instrumentation will modify exports